### PR TITLE
De-Reshape MatMul

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - *Introduced a transformation that removes Reshape operations surrounding MatMul operations if it is possible*

This change benefits LLMs -- it clears Scaled Dot Product patterns out of unnecessary Reshape operations. It makes it easier to try to match Scaled Dot Product uniformly across different models. Uniformed matching is not implemented in current PR. It will be added in future PR if proved to be beneficial.

Simplest case:
MicrosoftTeams-image

More complex example:
MicrosoftTeams-image (1)

### Tickets:
 - *CVS-112290*

### CODE REVIEW

File: src/common/transformations/include/transformations/symbolic_transformations/dereshape_matmul.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <openvino/pass/graph_rewrite.hpp>
```
[the remark this answers is not in the document]
Comment (jane-intel) [reply src/common/transformations/include/transformations/symbolic_transformations/dereshape_matmul.hpp:8]: Will change that together with major comments

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+bool batches_are_equal(const std::shared_ptr<ov::Node>& op_0, const std::shared_ptr<ov::Node>& op_1) {
+    auto input_0 = op_0->get_input_partial_shape(0);
+    auto input_1 = op_1->get_input_partial_shape(0);
+    auto output_0 = op_0->get_output_partial_shape(0);
+    auto output_1 = op_1->get_output_partial_shape(0);
+    return batches_are_equal(input_0, input_1, true) && batches_are_equal(output_0, output_1);
+}
+
+ov::Output<ov::Node> get_shape_from_sources(const ov::Output<ov::Node>& batch_dims_source,
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:65]: maybe we can combine 3 loops into 1?
```
size_t num_batch_dims = batch_dims_source.get_partial_shape().size() - 2;
size_t num_non_batch_dims = 2;

ov::NodeVector dims;
std::vector<size_t> non_constant_ids;

std::shared_ptr<Node> prev, curr;
for (size_t i = 0; i < num_batch_dims + num_non_batch_dims; ++i) {
    prev = curr;
    const auto& dims_source = i < num_batch_dims ? batch_dims_source : non_batch_dims_source;
    auto idx = i < num_batch_dims? i : non_batch_dims_source.get_partial_shape().size() - (i - num_batch_dims);
    auto current = ov::op::util::node_to_get_shape_value_of_indices_from_shape_source(dims_source,  {idx}, copy_rt_info_from);

    OPENVINO_SUPPRESS_DEPRECATED_START
     if (auto constant = ov::get_constant_from_source(node)) {
            OPENVINO_SUPPRESS_DEPRECATED_END
            if (ov::op::util::is_constant(prev)) {
                  curr = ov::op::util::make_try_fold<ov::op::v0::Concat>(ov::NodeVector{prev_node, curr_node}, 0);
                  dims.push_back(curr);
            } else {
                   if (i < num_batch_dims)
                          non_constant_ids.push_back(dims.size() - 1);
                   dims.push_back(prev);
            }
     } 
}
if (non_constant_ids.size() == 1) {
   // general case here if dims array contains Constant NonConstant Constant
   // NonConstant will be replaced with ov::op::v0::Constant::create(ov::element::i64, {1}, {-1});
   // so we need to call make_try_fold here twice 
   
```

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        OPENVINO_SUPPRESS_DEPRECATED_START
+        if (auto constant = ov::get_constant_from_source(node)) {
+            OPENVINO_SUPPRESS_DEPRECATED_END
+            node = constant;
+        } else {
+            non_constant_ids.push_back(i);
+        }
+        dims.push_back(node);
+    }
+    if (non_constant_ids.size() == 1) {
```
[the remark this answers is not in the document]
Comment (jane-intel) [reply src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:84]: yes

File: src/common/transformations/src/transformations/symbolic_transformations/utils.cpp [hunk 4]
```diff
@@ -32,3 +32,20 @@ bool ov::symbol::util::are_unique_and_equal_labels(const ov::TensorLabel& lhs, c
...
+    auto rhs_label = ov::DimensionTracker::get_label(rhs);
+    auto table_l = ov::DimensionTracker::get_table_of_equivalence(lhs);
+    auto table_r = ov::DimensionTracker::get_table_of_equivalence(rhs);
+    if (table_l)
+        labels_exist_and_equal = lhs_label != ov::no_label && table_l->are_equal(lhs, rhs);
+    else if (table_r)
+        labels_exist_and_equal = lhs_label != ov::no_label && table_r->are_equal(lhs, rhs);
+    else
+        labels_exist_and_equal = lhs_label != ov::no_label && lhs_label == rhs_label;
+    bool dims_are_static_and_equal = lhs.is_static() && lhs == rhs;
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/utils.cpp:49]: minor:
I believe it should check it first in this function
```
if (lhs.is_static() && lhs == rhs)
    return true;
```

File: src/common/transformations/include/transformations/symbolic_transformations/symbolic_optimizations.hpp [hunk 5]
```diff
@@ -48,3 +49,19 @@ class ov::pass::SymbolicPropagation : public ov::pass::ModelPass {
...
+ * @ingroup ie_transformation_common_api
+ * @brief Transformation requires equal labels on one input of Add and output of last Reshape in the pattern:
+ *      -> Add -> Reshape -[then or else input]-> Select -> Softmax -> Reshape ->
+ *
+ * If shape labels onn mentioned tensors are equal we proved that no broadcasting of this input was done for Add and
+ * for Select. Therefore, we can put the same labels on the output of Add and Select. This transformation helps
+ * propagate labels and will not be needed if we would use information on equality of products of input and output
+ * dimensions of Reshape operations
+ */
+class ov::pass::LabelResolvingThroughSelect : public ov::pass::MatcherPass {
```
Comment (itikhono) [inline src/common/transformations/include/transformations/symbolic_transformations/symbolic_optimizations.hpp:63]: please move it to the separate file

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+    auto rhs_bea_scalar = SCALAR_INPUT;
+    auto rhs_bea = pattern::wrap_type<op::util::BinaryElementwiseArithmetic>({rhs_reshape_or_concat, rhs_bea_scalar},
+                                                                             pattern::consumers_count(1));
+
+    auto rhs_bea_or_concat = std::make_shared<pattern::op::Or>(OutputVector{rhs_reshape_or_concat, rhs_bea});
+    // END: symmetrical patterns for MatMul inputs
+
+    auto matmul =
+        pattern::wrap_type<op::v0::MatMul>({lhs_bea_or_concat, rhs_bea_or_concat}, pattern::consumers_count(1));
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:227]: is it possible to use just 1 pattern, e. g lhs_bea_or_concat? and delete rhs_bea_or_concat [pattern
pattern::wrap_type<op::v0::MatMul>({lhs_bea_or_concat, lhs_bea_or_concat}

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+                                                  std::shared_ptr<ov::Node> concat_label,
+                                                  std::shared_ptr<ov::Node> bea_label,
+                                                  ov::Output<ov::Node> reshape_output,
+                                                  ov::Input<ov::Node> matmul_input,
+                                                  std::vector<ov::Node*>& nodes_for_revalidation) {
+    // Reshape -- [Concat] -- [BEA with scalar] -- > MatMul
+    auto original_reshape = reshape_output.get_node_shared_ptr();
+    if (vm.count(concat_label)) {
+        auto concat_node = ov::as_type_ptr<ov::op::v0::Concat>(vm.at(concat_label).get_node_shared_ptr());
+        OPENVINO_ASSERT(concat_node != nullptr,
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:135]: usually transformations do not use ASSERT 
As I understand, the main idea is to continue execution without applying the transformation, so maybe it's better to use `return false;` and stop the transformation

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                        "DeReshapeMatMul transformation matched operation which should be Concat -- but it is not");
+        auto rank = concat_node->get_output_partial_shape(0).rank().get_length();
+        auto axis = (concat_node->get_concatenation_axis() == (rank - 1)) ? -1 : -2;
+
+        auto idx_of_reshape_input = reshape_output == concat_node->input_value(0) ? 0 : 1;
+        auto idx_of_non_reshape_input = static_cast<size_t>(!idx_of_reshape_input);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:141]: better 
auto idx_of_non_reshape_input = 1 - idx_of_reshape_input;

File: src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+                                                     pm.at(matmul)->input(1),
+                                                     nodes_for_revalidation);
+
+        for (auto& node : nodes_for_revalidation)
+            node->validate_and_infer_types();
+
+        if (vm.count(add)) {
+            auto add_node = pm.at(add);
+            size_t matmul_port = (add_node->input_value(0) == vm.at(matmul) ? 0 : 1);
+            size_t non_matmul_port = static_cast<size_t>(!matmul_port);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/symbolic_transformations/dereshape_matmul.cpp:306]: better
auto non_matmul_port = 1 - matmul_port;

**Review by itikhono — APPROVED** [review]
in general, LGTM

**Pull request merged**